node/core/approval-voting/src/lib.rs

```diff
@@ -689,28 +788,47 @@ async fn run<C>(
 
 				actions
 			}
-			background_request = background_rx.next().fuse() => {
-				if let Some(req) = background_request {
-					handle_background_request(
-						&mut ctx,
-						&mut state,
-						&subsystem.metrics,
-						req,
-					).await?
-				} else {
-					Vec::new()
+			approval_state = currently_checking_set.next(&mut approvals_cache).fuse() => {
+				let mut actions = Vec::new();
+				let (
+					relay_block_hashes,
+					ApprovalState {
+						validator_index,
+						candidate_hash: _,
+						candidate_index,
+						approval_outcome,
+					}
+				) = approval_state;
+
+				if matches!(approval_outcome, ApprovalOutcome::Approved) {
+					let mut approvals: Vec<Action> = relay_block_hashes
+						.into_iter()
+						.map(|block_hash|
+							Action::IssueApproval(
+								ApprovalVoteRequest {
+									validator_index,
+									block_hash,
+									candidate_index,
```

Comment:
Here's the bug I mentioned before - the `candidate_index` is going to be wrongly specified for any block which didn't have work directly launched for it.